### PR TITLE
mgr/dashboard_v2: Data transform in k/v table

### PR DESCRIPTION
**Pull request by Devp00l**
The key value table can transform the input data into key value objects
now, you don't have to care about that anymore.

Also the first column of the data used in cd-table will be sorted if
not set manual through the "sorts" array.

Signed-off-by: Stephan Müller <smueller@suse.com>

### CODE REVIEW

**Review by LenzGr — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts [hunk 1]
```diff
@@ -42,6 +42,42 @@ export class TableKeyValueComponent implements OnInit {
   constructor() { }
 
   ngOnInit() {
+    if (!this.data) {
+      throw new Error('No data provided');
+    }
+    if (this.data.hasOwnProperty('length')) {
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts]: Could you replace this and the following conditions with lodash validations?

**Review by tspmelo — CHANGES_REQUESTED** [review]

**Comment by ricardoasmarques** [comment]
While testing this PR I found the following error:

screenshot from 2018-02-19 13-18-19

**Review by ricardoasmarques — CHANGES_REQUESTED** [review]
Found one JS error while testing this PR

**Comment by LenzGr** [comment]
_Found one JS error while testing this PR_

Thanks for spotting this, @ricardoasmarques. Is there a way to test for these automatically? Or is there a way to spot such errors manually before submitting a PR?

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts (hunk 1 above)
Comment (Devp00l) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts]: done

**Comment by Devp00l** [comment]
@ricardoasmarques @LenzGr : The JS-Error is finally fixed. The cause of this error was implemented before but somehow not triggered, that's why it was so hard to find the error.

**Review by votdev — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/rgw/rgw-daemon-details/rgw-daemon-details.component.ts [hunk 2]
```diff
@@ -1,17 +1,17 @@
 import { Component, Input, OnInit } from '@angular/core';
 
-import * as _ from 'lodash';
-
 import { RgwDaemonService } from '../services/rgw-daemon.service';
 
+import * as _ from 'lodash';
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/rgw/rgw-daemon-details/rgw-daemon-details.component.ts]: please revert this change.

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts [hunk 3]
```diff
@@ -49,15 +39,58 @@ export class TableKeyValueComponent implements OnInit {
...
+    this.useData();
+  }
+
+  useData() {
+    let temp = [];
+    if (!this.data) {
+      return; // Wait for data
+    } else if (_.isArray(this.data)) {
+      const first = this.data[0];
+      if (_.isObject(first) && _.has(first, 'key') && _.has(first, 'value')) {
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts]: we should replace `_.isObject` with `_.isPlainObject`.
`_.isObject` will return true if you test an array.

**Review by tspmelo — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts (hunk 3 above)
Comment (Devp00l) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/shared/datatable/table-key-value/table-key-value.component.ts]: done

File: src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/rgw/rgw-daemon-details/rgw-daemon-details.component.ts (hunk 2 above)
Comment (Devp00l) [reply src/pybind/mgr/dashboard_v2/frontend/src/app/ceph/rgw/rgw-daemon-details/rgw-daemon-details.component.ts]: done

**Review by ricardoasmarques — DISMISSED** [review]
lgtm

**Review by tspmelo — APPROVED** [review]
